src/d3a/d3a_core/sim_results/plotly_graph.py

```diff
@@ -0,0 +1,584 @@
+"""
+Copyright 2018 Grid Singularity
+This file is part of D3A.
+
+This program is free software: you can redistribute it and/or modify
+it under the terms of the GNU General Public License as published by
+the Free Software Foundation, either version 3 of the License, or
+(at your option) any later version.
+
+This program is distributed in the hope that it will be useful,
+but WITHOUT ANY WARRANTY; without even the implied warranty of
+MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+GNU General Public License for more details.
+
+You should have received a copy of the GNU General Public License
+along with this program.  If not, see <http://www.gnu.org/licenses/>.
+"""
+import logging
+import plotly as py
+import plotly.graph_objs as go
+import pendulum
+from sortedcontainers import SortedDict
+
+from d3a.constants import TIME_ZONE
+from d3a.models.strategy.storage import StorageStrategy
+from d3a.models.strategy.load_hours import LoadHoursStrategy
+from d3a.models.strategy.pv import PVStrategy
+from d3a import limit_float_precision
+
+_log = logging.getLogger(__name__)
+
+ENERGY_BUYER_SIGN_PLOTS = 1
+ENERGY_SELLER_SIGN_PLOTS = -1 * ENERGY_BUYER_SIGN_PLOTS
+
+alternative_pricing_subdirs = {
+    0: "d3a_pricing",
+    1: "no_scheme_pricing",
+    2: "feed_in_tariff_pricing",
+    3: "net_metering_pricing"
+}
+
+EXPORT_DEVICE_VARIABLES = ["trade_energy_kWh", "pv_production_kWh", "trade_price_eur",
+                           "soc_history_%", "load_profile_kWh"]
+
+green = 'rgba(20,150,20, alpha)'
+purple = 'rgb(156, 110, 177, alpha)'
+blue = 'rgba(0,0,200,alpha)'
+
+DEVICE_PLOT_COLORS = {"trade_energy_kWh": purple,
+                      "pv_production_kWh": green,
+                      "load_profile_kWh": green,
+                      "soc_history_%": green,
+                      "trade_price_eur": blue}
+
+DEVICE_YAXIS = {"trade_energy_kWh": 'Demand/Traded [kWh]',
+                "pv_production_kWh": 'PV Production [kWh]',
+                "load_profile_kWh": 'Load Profile [kWh]',
+                "soc_history_%": 'State of Charge [%]',
+                "trade_price_eur": 'Energy Rate [EUR/kWh]'}
+
+
+def _invert(inlist: list): return [-1 * l for l in inlist]
+
+
+def _get_color(key, alpha):
+    return DEVICE_PLOT_COLORS[key].replace("alpha", str(alpha))
+
+
+class PlotlyGraph:
+    def __init__(self, dataset: dict, key: str):
+        self.key = key
+        self.dataset = dataset
+        self.umHours = dict()
+        self.rate = list()
+        self.energy = list()
+        self.trade_history = dict()
+
+    @staticmethod
+    def _line_plot(curve_point, time, supply):
+        graph_obj = PlotlyGraph(curve_point, time)
```

Comment:
This is actually weird, since the object of the class is instantiated inside the same class. To my eyes this looks very error-prone, and I would suggest to move this method out of the class, in the same module. It is a static method anyways therefore it does not need any member of the class.

Reply:
You are right, will fix this